### PR TITLE
DDCNL-7207

### PR DESCRIPTION
**Pull request by Nj147**
- [ ] Unit tests passing
- [ ] Coverage reached
- [ ] Acceptance tests passing
- [ ] Reviewed

### CODE REVIEW

File: project/AppDependencies.scala [hunk 1]
```diff
@@ -5,43 +5,51 @@ private object AppDependencies {
...
-          "com.typesafe.play"      %% "play-test"          % PlayVersion.current % scope,
-          "org.scalacheck"         %% "scalacheck"         % "1.14.3"            % scope,
-          "uk.gov.hmrc"            %% s"bootstrap-test-$playVersion"             % "5.3.0"
-        )
-      }.test
-  }
-
-  def apply() = compile ++ Test()
+  val test: Seq[ModuleID] = Seq(
+          "org.scalatestplus.play" %% "scalatestplus-play"            % "5.1.0",
```
Comment (Christian-QA) [inline project/AppDependencies.scala:23]: `scalatestplus-play` should be included in `bootstrap-test`

Comment (Nj147) [reply project/AppDependencies.scala]: Pushed up this change

File: app/uk/gov/hmrc/tai/service/AuditService.scala [hunk 2]
```diff
@@ -51,12 +50,14 @@ class AuditService @Inject() (
 
   def createAndSendAuditEvent(eventName: String, details: Map[String, String])(implicit
     hc: HeaderCarrier,
-    request: Request[AnyContent]
+    request: Request[AnyContent],
+    executionContext: ExecutionContext
```
Comment (pascal22p) [inline app/uk/gov/hmrc/tai/service/AuditService.scala:54]: Inject the execution context in the class instead

File: app/uk/gov/hmrc/tai/service/AuditService.scala [hunk 3]
```diff
@@ -51,12 +50,14 @@ class AuditService @Inject() (
...
   ): Future[AuditResult] =
     createAndSendAuditEvent(eventName, fetchPath(request), details)
 
   def createAndSendAuditEvent(eventName: String, path: String, details: Map[String, String])(implicit
-    hc: HeaderCarrier
+    hc: HeaderCarrier,
+    executionContext: ExecutionContext
```
Comment (pascal22p) [inline app/uk/gov/hmrc/tai/service/AuditService.scala:60]: same

File: app/uk/gov/hmrc/tai/service/AuditService.scala [hunk 4]
```diff
@@ -74,7 +75,7 @@ class AuditService @Inject() (
     numberOfEmployments: Seq[Employment],
     numberOfTaxCodeIncomes: Seq[TaxCodeIncome],
     isJrsTileShown: Boolean
-  )(implicit hc: HeaderCarrier, request: Request[_]): Future[AuditResult] = {
+  )(implicit hc: HeaderCarrier, request: Request[_], executionContext: ExecutionContext): Future[AuditResult] = {
```
Comment (pascal22p) [inline app/uk/gov/hmrc/tai/service/AuditService.scala:78]: same

File: app/uk/gov/hmrc/tai/service/AuditService.scala [hunk 5]
```diff
@@ -92,7 +93,7 @@ class AuditService @Inject() (
     fuelBenefitDate: Option[String],
     isSuccessful: Boolean,
     path: String
-  )(implicit hc: HeaderCarrier, request: Request[_]): Future[AuditResult] = {
+  )(implicit hc: HeaderCarrier, request: Request[_], executionContext: ExecutionContext): Future[AuditResult] = {
```
Comment (pascal22p) [inline app/uk/gov/hmrc/tai/service/AuditService.scala:96]: same

File: app/uk/gov/hmrc/tai/service/AuditService.scala [hunk 6]
```diff
@@ -107,7 +108,8 @@ class AuditService @Inject() (
 
   def sendAuditEventAndGetRedirectUri(nino: Nino, iformName: String)(implicit
     hc: HeaderCarrier,
-    request: Request[AnyContent]
+    request: Request[AnyContent],
+    executionContext: ExecutionContext
```
Comment (pascal22p) [inline app/uk/gov/hmrc/tai/service/AuditService.scala:112]: same

File: app/uk/gov/hmrc/tai/service/CodingComponentService.scala [hunk 7]
```diff
@@ -24,18 +24,22 @@ import uk.gov.hmrc.tai.model.domain.TaxFreeAmountComparison
...
 
 class CodingComponentService @Inject() (
   taxAccountConnector: TaxAccountConnector,
   taxFreeAmountComparisonConnector: TaxFreeAmountComparisonConnector
 ) {
 
-  def taxFreeAmountComponents(nino: Nino, year: TaxYear)(implicit hc: HeaderCarrier): Future[Seq[CodingComponent]] =
+  def taxFreeAmountComponents(nino: Nino, year: TaxYear)(implicit
+    hc: HeaderCarrier,
+    executionContext: ExecutionContext
```
Comment (pascal22p) [inline app/uk/gov/hmrc/tai/service/CodingComponentService.scala:36]: inject the execution context in the class instead

File: app/uk/gov/hmrc/tai/service/CodingComponentService.scala [hunk 8]
```diff
@@ -24,18 +24,22 @@ import uk.gov.hmrc.tai.model.domain.TaxFreeAmountComparison
...
+  ): Future[Seq[CodingComponent]] =
     taxAccountConnector.codingComponents(nino, year).map(filterOutZeroAmountsComponents)
 
-  def taxFreeAmountComparison(nino: Nino)(implicit hc: HeaderCarrier): Future[TaxFreeAmountComparison] =
+  def taxFreeAmountComparison(
+    nino: Nino
+  )(implicit hc: HeaderCarrier, executionContext: ExecutionContext): Future[TaxFreeAmountComparison] =
```
Comment (pascal22p) [inline app/uk/gov/hmrc/tai/service/CodingComponentService.scala:42]: same here

File: app/uk/gov/hmrc/tai/service/TaxCodeChangeService.scala [hunk 9]
```diff
@@ -22,22 +22,22 @@ import play.api.Logging
...
 
 class TaxCodeChangeService @Inject() (taxCodeChangeConnector: TaxCodeChangeConnector) extends Logging {
 
   def taxCodeChange(nino: Nino)(implicit hc: HeaderCarrier): Future[TaxCodeChange] =
     taxCodeChangeConnector.taxCodeChange(nino)
 
-  def hasTaxCodeChanged(nino: Nino)(implicit hc: HeaderCarrier): Future[Either[TaxCodeError, HasTaxCodeChanged]] = {
+  def hasTaxCodeChanged(
+    nino: Nino
+  )(implicit hc: HeaderCarrier, executionContext: ExecutionContext): Future[Either[TaxCodeError, HasTaxCodeChanged]] = {
```
Comment (pascal22p) [inline app/uk/gov/hmrc/tai/service/TaxCodeChangeService.scala:40]: inject execution context in the class instead

File: build.sbt [hunk 10]
```diff
@@ -95,22 +94,17 @@ lazy val microservice = Project(appName, file("."))
     // only compress files generated by concat
     uglify / includeFilter := GlobFilter("tai-*.js")
   )
-  .settings(
```
[the remark this answers is not in the document]
Comment (pascal22p) [reply build.sbt:98]: I agree with Christian. Raise a ticket to tacke the warnings

**Review by pascal22p — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/tai/connectors/HttpHandler.scala [hunk 11]
```diff
@@ -16,19 +16,39 @@
...
+      case Left(error) if error.statusCode >= 499 || error.statusCode == TOO_MANY_REQUESTS =>
+        logger.error(error.message)
+        Left(error)
+      case Left(error) =>
+        logger.error(error.message, error)
+        Left(error)
+    } recover { case exception: HttpException =>
+      logger.error(exception.message)
+      Left(UpstreamErrorResponse(exception.message, 502, 502))
+    })
```
[the remark this answers is not in the document]
Comment (Nj147) [reply app/uk/gov/hmrc/tai/connectors/HttpHandler.scala]: Yeah think it was brought in from a stash from the messagefrontend ticket, removed now

**Review by pascal22p — APPROVED** [review]

**Pull request merged**